### PR TITLE
Add support for Razer Blackwidow v3 Pro wireless

### LINKED ISSUE #1279: Support for BlackWidow V3 Pro [1532:025c]
### Device Info

* Name: BlackWidow V3 Pro
* Product Number: RZ03-03530200-R3U1
* Model Number: RZ03-0353

### Device Information

<details>
  <summary> lsusb -d 1532: -v </summary>

  ```
Bus 001 Device 003: ID 1532:025c Razer USA, Ltd Razer BlackWidow V3 Pro
Couldn't open device, some information will be missing
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x025c 
  bcdDevice            1.00
  iManufacturer           1 
  iProduct                2 
  iSerial                 3 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0086
    bNumInterfaces          5
    bConfigurationValue     1
    iConfiguration          4 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              100mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     179
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      94
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        4
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     325
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x85  EP 5 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        3
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     348
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x84  EP 4 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
  ```
</details>
    
<details>
  <summary>ls -lah /dev/input/by-id/</summary>

  ```
total 0
drwxr-xr-x 2 root root 240 Oct  3 20:34 .
drwxr-xr-x 4 root root 800 Oct  3 20:34 ..
lrwxrwxrwx 1 root root   9 Oct  1 17:28 usb-Logitech_USB_Receiver-if02-event-kbd -> ../event6
lrwxrwxrwx 1 root root   9 Oct  1 17:28 usb-Logitech_USB_Receiver-if02-event-mouse -> ../event4
lrwxrwxrwx 1 root root   9 Oct  1 17:28 usb-Logitech_USB_Receiver-if02-mouse -> ../mouse2
lrwxrwxrwx 1 root root  10 Oct  3 20:34 usb-Razer_Razer_BlackWidow_V3_Pro_000000000000-event-if01 -> ../event27
lrwxrwxrwx 1 root root  10 Oct  3 20:34 usb-Razer_Razer_BlackWidow_V3_Pro_000000000000-event-if03 -> ../event29
lrwxrwxrwx 1 root root  10 Oct  3 20:34 usb-Razer_Razer_BlackWidow_V3_Pro_000000000000-event-kbd -> ../event23
lrwxrwxrwx 1 root root  10 Oct  3 20:34 usb-Razer_Razer_BlackWidow_V3_Pro_000000000000-if01-event-kbd -> ../event24
lrwxrwxrwx 1 root root  10 Oct  3 20:34 usb-Razer_Razer_BlackWidow_V3_Pro_000000000000-if02-event-mouse -> ../event28
lrwxrwxrwx 1 root root   9 Oct  3 20:34 usb-Razer_Razer_BlackWidow_V3_Pro_000000000000-if02-mouse -> ../mouse4
lrwxrwxrwx 1 root root   9 Oct  1 17:28 usb-SteelSeries_Arctis_Pro_Wireless-event-if05 -> ../event7
  ```
</details>

### Packet Captures

<Please see here as we probably need packet captures>
<https://github.com/openrazer/openrazer/wiki/Reverse-Engineering-USB-Protocol>

### Synapse Screenshots
<Screenshots of all available options as seen on Razer Synapse>
(Coming soon)

### PR DESCRIPTION
**Pull request by nicfix**
Added partial support for BlackWidow V3 Pro wireless

What works:
- The device is detected (tested on polychromatic)
- All the effects work apart from ripple (disabled)

What doesn't work:
- Ripple effect

What works meh:
- Battery level, I still get several notifications of empty battery, same behavior as my basilisk ultimate though.

This PR is heavily inspired by the one from @GeoDoX  (https://github.com/openrazer/openrazer/pull/1623), I thought about forking from that branch but rebasing @GeoDoX's branch proved more difficult than starting from scratch.

Partially fixes #1279

### CODE REVIEW

**Comment by andriinuts** [comment]
@z3ntu please have a look at this PR as the previous one was stuck

**Comment by nicfix** [comment]
@z3ntu any news for this PR?

File: driver/razerkbd_driver.c [hunk 1]
```diff
@@ -352,9 +357,13 @@ static void razer_set_device_mode(struct usb_device *usb_dev, unsigned char mode
     case USB_DEVICE_ID_RAZER_BLACKWIDOW_V3_MINI:
         report.transaction_id.id = 0x1F;
         break;
+    case USB_DEVICE_ID_RAZER_BLACKWIDOW_V3_PRO_WIRELESS:
     case USB_DEVICE_ID_RAZER_BLACKWIDOW_V3_MINI_WIRELESS:
         report.transaction_id.id = 0x9F;
         break;
+    default:
+        report.transaction_id.id = 0x1f;
+        break;
```
Comment (z3ntu) [inline driver/razerkbd_driver.c]: Why is this block added?

File: driver/razerkbd_driver.c [hunk 2]
```diff
@@ -1533,7 +1549,11 @@ static ssize_t razer_attr_write_matrix_effect_reactive(struct device *dev, struc
     case USB_DEVICE_ID_RAZER_BLACKWIDOW_V3_PRO_WIRED:
         report = razer_chroma_extended_matrix_effect_reactive(VARSTORE, BACKLIGHT_LED, speed, (struct razer_rgb*)&buf[1]);
         break;
-
+    case USB_DEVICE_ID_RAZER_BLACKWIDOW_V3_PRO_WIRELESS:
+        razer_set_device_mode(usb_dev, 0x00, 0x00);
```
Comment (z3ntu) [inline driver/razerkbd_driver.c:1618]: If the daemon likes to switch the device to driver mode (which I didn't check now), then maybe that should be removed if reactive effect doesn't work with it.
Overwriting the device mode here isn't great at all...

File: driver/razerkbd_driver.c [hunk 3]
```diff
@@ -1699,7 +1718,7 @@ static ssize_t razer_attr_write_matrix_effect_static(struct device *dev, struct
         report.transaction_id.id = 0x1F;
         razer_send_payload(usb_dev, &report);
         break;
-
```
Comment (z3ntu) [inline driver/razerkbd_driver.c:1772]: Please revert all these unrelated whitespace changes in this file

File: daemon/openrazer_daemon/hardware/keyboards.py [hunk 4]
```diff
@@ -382,11 +382,12 @@ class RazerBlackWidowV3ProWired(_RippleKeyboard):
                'set_charge_effect', 'set_charge_colour', 'get_battery', 'is_charging', 'set_idle_time', 'set_low_battery_threshold']
 
     DEVICE_IMAGE = "https://dl.razerzone.com/src/3809-1-EN-v1.png"
+    DEVICE_NAME = "Razer BlackWidow V3 Pro (Wired)"
```
Comment (z3ntu) [inline daemon/openrazer_daemon/hardware/keyboards.py]: Please just follow the style from other devices and don't invent a way of doing things

File: daemon/openrazer_daemon/hardware/keyboards.py [hunk 5]
```diff
@@ -397,6 +398,22 @@ def _close(self):
...
+    Class for the Razer BlackWidow v3 Pro 2.4 Ghz Wireless
+    """
+    USB_PID = 0x025C
+    DEVICE_NAME = "Razer BlackWidow V3 Pro (Wireless)"
+    METHODS = ['get_device_type_keyboard', 'set_wave_effect', 'set_static_effect', 'set_spectrum_effect',
+               'set_reactive_effect', 'set_none_effect', 'set_breath_random_effect', 'set_breath_single_effect', 'set_breath_dual_effect',
+               'set_custom_effect', 'set_key_row', 'get_game_mode', 'set_game_mode', 'get_macro_mode', 'set_macro_mode',
+               'get_macro_effect', 'set_macro_effect', 'get_macros', 'delete_macro', 'add_macro',
+               'set_starlight_random_effect', 'set_starlight_single_effect', 'set_starlight_dual_effect',
+               # 'set_ripple_effect', 'set_ripple_effect_random_colour',
```
Comment (z3ntu) [inline daemon/openrazer_daemon/hardware/keyboards.py:411]: Remove the comment if it doesn't work?

**Comment by stat1x** [comment]
Would be cool if it gets merged.
Thanks.

**Comment by YoMamasNerd** [comment]
Nice, I`d love to test this. 
I`ve just got this keyboard.

Happy Holidays

**Comment by z3ntu** [comment]
@andriinuts @YoMamasNerd Does this PR work for you? See https://github.com/openrazer/openrazer/wiki/Building-a-package

@nicfix There's still some open comments, could you resolve them?

**Comment by YoMamasNerd** [comment]
I'll try it this weekend probably.
Thank you for your response @z3ntu 🙂

**Comment by YoMamasNerd** [comment]
@z3ntu I may need some further help to get this working. Do you have a discord or something?
The Keyboard is working in wireless mode. Sleep mode settings do not work. 

```
Traceback (most recent call last):
  File ’/usr/lib/python3.10/site-packages/polychromatic/controller/devices.py’, line 429, in _slider_dropped
    option.apply(slider.value())
  File ’/usr/lib/python3.10/site-packages/polychromatic/backends/openrazer.py’, line 1339, in apply
    self._rdevice.set_idle_time(int(new_value) * 60)
  File ’/usr/lib/python3.10/site-packages/openrazer/client/devices/__init__.py’, line 460, in set_idle_time
    self._dbus_interfaces[’power’].setIdleTime(idle_time)
  File ’/usr/lib/python3.10/site-packages/dbus/proxies.py’, line 141, in __call__
    return self._connection.call_blocking(self._named_service,
  File ’/usr/lib/python3.10/site-packages/dbus/connection.py’, line 652, in call_blocking
    reply_message = self.send_message_with_reply_and_block(
dbus.exceptions.DBusException: org.freedesktop.DBus.Python.PermissionError: Traceback (most recent call last):
  File ’/usr/lib/python3.10/site-packages/dbus/service.py’, line 715, in _message_cb
    retval = candidate_method(self, *args, **keywords)
  File ’/usr/lib/python3.10/site-packages/openrazer_daemon/dbus_services/dbus_methods/mamba.py’, line 51, in set_idle_time
    with open(driver_path, ’w’) as driver_file:
PermissionError: [Errno 13] Permission denied: ’/sys/devices/pci0000:00/0000:00:01.3/0000:02:00.0/usb1/1-3/1-3:1.2/0003:1532:025C.0006/device_idle_time’

```

Battery reporting is working with the keyboard.

And my mouse (Deathadder v3 Pro) is not getting recognized anymore. May be a result of being so far behind the official repo.

**Comment by z3ntu** [comment]
@YoMamasNerd Telegram or Matrix (or IRC if necessary): https://github.com/openrazer/openrazer/wiki/Community-chat

Regarding this error, looks like the device declares it supports `set_idle_time` but `device_idle_time` is not implemented in kernel driver so it fails. This should be corrected (how depends on whether the firmware supports this command or not)

File: driver/razerkbd_driver.c (hunk 1 above)
Comment (z3ntu) [reply driver/razerkbd_driver.c]: removed it for now, still keeping this "conversation" open, maybe it was super important for something

File: driver/razerkbd_driver.c (hunk 3 above)
Comment (z3ntu) [reply driver/razerkbd_driver.c:1772]: done it myself

**Comment by YoMamasNerd** [comment]
Things that work

- Effects
- changing brightness 
- updating battery level
- low power mode settings

Things that does not work

- setting "Sleep mode after"
- updating charging status without reloading Polychromatic
- Battery warning shows Labeled tag

**Comment by z3ntu** [comment]
Has there been any progress on the non-working items?

**Comment by YoMamasNerd** [comment]
Not yet. I've been too busy to fix the issues since they are not breaking anything for me.
I might be able to look into resolving the issues this weekend.

**Comment by z3ntu** [comment]
Closing in favor of #2056
